### PR TITLE
Histogram Rollout - Algorithms 98-102

### LINKED ISSUE #17752: HistogramData Rollout: Algorithms 98-102
HistogramData rollout: algorithms 98-102

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:
```
Framework/DataHandling/src/GroupDetectors.cpp
Framework/DataHandling/src/ImggAggregateWavelengths.cpp
Framework/DataHandling/src/LoadAscii2.cpp
Framework/DataHandling/src/LoadAscii.cpp
Framework/DataHandling/src/LoadCalFile.cpp
```

### PR DESCRIPTION
**Pull request by DavidFair**
This PR requires  #17788 to be merged first.

HistogramData rollout: algorithms 98-102
## Description of work.

This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

```
Framework/DataHandling/src/GroupDetectors.cpp
Framework/DataHandling/src/ImggAggregateWavelengths.cpp
Framework/DataHandling/src/LoadAscii2.cpp
Framework/DataHandling/src/LoadAscii.cpp
Framework/DataHandling/src/LoadCalFile.cpp
```

Performance tests were added for:

```
LoadAscii2
LoadCalFile
```

The following algorithms have improved performance:
None
## To test.

Code review.

Fixes #17752 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/DataHandling/src/ImggAggregateWavelengths.cpp [hunk 1]
```diff
@@ -778,10 +778,10 @@ void ImggAggregateWavelengths::aggImage(API::MatrixWorkspace_sptr accum,
...
 
   for (size_t row = 0; row < sizeY; row++) {
-    auto &dataY = accum->dataY(row);
-    const auto &dataYIn = toAdd->readY(row);
-    std::transform(dataY.begin(), dataY.end(), dataYIn.cbegin(), dataY.begin(),
-                   std::plus<double>());
+    auto &outputY = accum->mutableY(row);
+    const auto &yValsInput = toAdd->y(row);
+    std::transform(outputY.begin(), outputY.end(), yValsInput.cbegin(),
+                   outputY.begin(), std::plus<double>());
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/ImggAggregateWavelengths.cpp]: Isn't this just `outputY->mutableY(row) += toAdd->y(row);`?

File: Framework/DataHandling/src/LoadAscii.cpp [hunk 2]
```diff
@@ -199,11 +199,15 @@ API::Workspace_sptr LoadAscii::readData(std::ifstream &file) const {
     }
 
     for (size_t i = 0; i < numSpectra; ++i) {
-      spectra[i].dataX().push_back(values[0]);
-      spectra[i].dataY().push_back(values[i * 2 + 1]);
+      auto resizedHistogram = spectra[i].histogram();
+      const size_t newPosition = resizedHistogram.x().size();
+      resizedHistogram.resize(newPosition + 1);
+      resizedHistogram.mutableX()[newPosition] = values[0];
+      resizedHistogram.mutableY()[newPosition] = values[i * 2 + 1];
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadAscii.cpp]: There is also `back()` (and `front()`) for these types, just like for `std::vector`.

File: Framework/DataHandling/src/LoadAscii.cpp [hunk 3]
```diff
@@ -199,11 +199,15 @@ API::Workspace_sptr LoadAscii::readData(std::ifstream &file) const {
     }
 
     for (size_t i = 0; i < numSpectra; ++i) {
-      spectra[i].dataX().push_back(values[0]);
-      spectra[i].dataY().push_back(values[i * 2 + 1]);
+      auto resizedHistogram = spectra[i].histogram();
+      const size_t newPosition = resizedHistogram.x().size();
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadAscii.cpp]: Careful here. The size of a `Histogram` (and what `resize` takes is the number of *bins* (or values). So if you take the X length your resize below is not what you want in the case of X as bin-edges.

I think Anton moved `Histogram::size()` to `public`, so once his PR is merged you can use that here, that would be safer an cleaner, i.e., you can then just do `hist.resize(hist.size() + 1)`.

File: Framework/DataHandling/src/LoadAscii2.cpp [hunk 4]
```diff
@@ -420,37 +421,44 @@ void LoadAscii2::processHeader(std::ifstream &file) {
...
 * @param[in] columns : the columns of values in the current line of data
 */
 void LoadAscii2::addToCurrentSpectra(std::list<std::string> &columns) {
   std::vector<double> values(m_baseCols, 0.);
   m_spectraStart = false;
   fillInputValues(values, columns);
   // add X and Y
-  m_curSpectra->dataX().push_back(values[0]);
-  m_curSpectra->dataY().push_back(values[1]);
+  const size_t newPosition = m_curSpectra->x().size(); // For 0 based index
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadAscii2.cpp]: See comment above.

File: Framework/DataHandling/test/LoadAscii2Test.h [hunk 5]
```diff
@@ -542,4 +542,90 @@ class LoadAscii2Test : public CxxTest::TestSuite {
...
+    for (int i = 0; i < numVecs; i++) {
+      auto &X = wsToSave->mutableX(i);
+      auto &Y = wsToSave->mutableY(i);
+      auto &E = wsToSave->mutableE(i);
+      for (int j = 0; j < xyLen; j++) {
+        X[j] = 1.5 * j / 0.9;
+        Y[j] = (i + 1) * (2. + 4. * X[j]);
+        E[j] = 1.;
+      }
+    }
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadAscii2Test.h]: No need to init this with fancy values for a performance test, all default should be fine in this case, I think.

**Review by SimonHeybrock — COMMENTED** [review]
Looks good apart from minor comments:

**Comment by DavidFair** [comment]
I've merged the branch from PR  #17788 in so that needs to me merged before this.

**Comment by SimonHeybrock** [comment]
Latest changes look good :+1: . Merge #17788 then :shipit:.

**Pull request merged**